File: reports_generator_node/parser.js
const fs = require('fs')
const path = require('path')
const { promisify } = require('util')

const readFileAsync = promisify(fs.readFile)

async function parserFile(path) {
  const content = await readFileAsync(path, 'utf-8')
  const lines = content.split(/\r?\n/)
  const parser = lines.map((line) => {
    const [id, food, price] = line.trim().split(',')
    return { id, food, price: Number(price) }
  })
  return parser
}

async function parserFiles(paths) {
  const promises = paths.map((e) => parserFile(path.join('..', 'reports', e)))
  return (await Promise.all(promises)).flat()
}

module.exports = { parserFile, parserFiles }

File: gen_report_elixir/lib/parser.ex
defmodule GenReportElixir.Parser do
  def parser_file(filename) do
    "../reports/#{filename}"
    |> File.stream!()
    |> Stream.map(fn line -> parser_line(line) end)
  end

  defp parser_line(line) do
    line
    |> String.trim()
    |> String.split(",")
    |> List.update_at(0, &String.downcase/1)
    |> List.update_at(1, &String.to_integer/1)
    |> List.update_at(2, &String.to_integer/1)
    |> List.update_at(3, &String.to_integer/1)
    |> List.update_at(4, &String.to_integer/1)
  end
end
